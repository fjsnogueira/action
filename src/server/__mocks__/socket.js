const socket = {
  emit: jest.fn(),
  setAuthToken: jest.fn()
};

export default socket;
