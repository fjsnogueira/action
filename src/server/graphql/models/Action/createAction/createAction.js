import getRethink from 'server/database/rethinkDriver';
import {ActionInput} from '../actionSchema';
import {
  GraphQLNonNull,
  GraphQLBoolean,
} from 'graphql';
import {requireSUOrTeamMember, requireWebsocket} from 'server/utils/authorization';
import {handleSchemaErrors} from 'server/utils/utils';
import createActionValidation from './createActionValidation';

export default {
  type: GraphQLBoolean,
  description: 'Create a new action, triggering a CreateCard for other viewers',
  args: {
    newAction: {
      type: new GraphQLNonNull(ActionInput),
      description: 'The new action including an id, teamMemberId, [agendaId], and sortOrder '
    }
  },
  async resolve(source, {newAction}, {authToken, socket}) {
    const r = getRethink();

    // AUTH
    const {id} = newAction;
    // format of id is teamId::taskIdPart
    const [teamId] = id.split('::');
    requireSUOrTeamMember(authToken, teamId);
    requireWebsocket(socket);

    // VALIDATION
    const schema = createActionValidation(authToken.tms);
    const {errors, data: validNewAction} = schema(newAction);
    handleSchemaErrors(errors);

    // RESOLUTION
    const now = new Date();
    const [userId] = validNewAction.teamMemberId.split('::');
    const action = {
      ...validNewAction,
      userId,
      createdAt: now,
      createdBy: authToken.sub,
      updatedAt: now,
      isComplete: false
    };
    await r.table('Action').insert(action);
  }
};
