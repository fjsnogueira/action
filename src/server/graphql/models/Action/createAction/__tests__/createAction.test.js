import getRethink from 'server/database/rethinkDriver';
import mockAuthToken from 'server/__tests__/setup/mockAuthToken';
import MockDate from 'mockdate';
import {__now} from 'server/__tests__/setup/mockTimes';
import fetchAndSerialize from 'server/__tests__/utils/fetchAndSerialize';
import DynamicSerializer from 'dynamic-serializer';
import MockDB from 'server/__tests__/setup/MockDB';
import expectAsyncToThrow from 'server/__tests__/utils/expectAsyncToThrow';
import socket from 'server/__mocks__/socket';
import createAction from 'server/graphql/models/Action/createAction/createAction';
import shortid from 'shortid';

MockDate.set(__now);
console.error = jest.fn();

describe('createAction', () => {
  test('creates an action from a meeting', async() => {
    // SETUP
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const {agendaItem, user, team, teamMember} = await mockDB.init()
      .newAgendaItem();

    // TEST
    const teamId = team[0].id;
    const teamMemberId = teamMember[7].id;
    const agendaId = agendaItem[0].id;
    const authToken = mockAuthToken(user[7]);
    const newAction = {
      id: `${teamId}::${shortid.generate()}`,
      teamMemberId,
      sortOrder: 0,
      agendaId
    };
    await createAction.resolve(undefined, {newAction}, {authToken, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      action: r.table('Action').get(newAction.id),
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
  });

  test('creates an action from the team dash', async() => {
    // SETUP
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const {user, team, teamMember} = await mockDB.init();

    // TEST
    const teamId = team[0].id;
    const teamMemberId = teamMember[7].id;
    const authToken = mockAuthToken(user[7]);
    const newAction = {
      id: `${teamId}::${shortid.generate()}`,
      teamMemberId,
      sortOrder: 0,
    };
    await createAction.resolve(undefined, {newAction}, {authToken, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      action: r.table('Action').get(newAction.id),
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
  });

  test('throws when no websocket is present', async() => {
    const authToken = {};
    await expectAsyncToThrow(createAction.resolve(undefined, {newAction: {}}, {authToken}));
  });

  test('throw when the caller is not a team member', async() => {
    // SETUP
    const mockDB = new MockDB();
    const {teamMember, user} = await mockDB.init();

    // TEST
    const authToken = mockAuthToken(user[1]);
    const teamMemberId = teamMember[7].id;
    const newAction = {
      id: `foo::${shortid.generate()}`,
      teamMemberId,
      sortOrder: 0,
    };
    await expectAsyncToThrow(createAction.resolve(undefined, {newAction}, {authToken, socket}), [mockDB.context.team.id]);
  });
});
