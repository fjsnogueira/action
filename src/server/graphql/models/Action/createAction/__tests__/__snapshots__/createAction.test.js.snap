// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createAction creates an action from a meeting 1`] = `
Object {
  "action": Object {
    "agendaId": 1,
    "createdAt": 2017-01-01T08:00:00.000Z,
    "createdBy": 2,
    "id": 0,
    "isComplete": false,
    "sortOrder": 0,
    "teamMemberId": 3,
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userId": 2,
  },
}
`;

exports[`createAction creates an action from the team dash 1`] = `
Object {
  "action": Object {
    "createdAt": 2017-01-01T08:00:00.000Z,
    "createdBy": 1,
    "id": 0,
    "isComplete": false,
    "sortOrder": 0,
    "teamMemberId": 2,
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userId": 1,
  },
}
`;

exports[`createAction throw when the caller is not a team member 1`] = `"{\\"_error\\":\\"You do not have access to team foo\\"}"`;

exports[`createAction throws when no websocket is present 1`] = `"Cannot read property 'split' of undefined"`;
